### PR TITLE
feature: supported OpenSSL 1.1.0

### PR DESCRIPTION
**Pull request by spacewander**
`EVP_CIPHER_CTX` is opaque after OpenSSL 1.1.0.
Replace `EVP_CIPHER_CTX_init` and `EVP_CIPHER_CTX_cleanup` with
`EVP_CIPHER_CTX_new` and `EVP_CIPHER_CTX_free`.

### CODE REVIEW

File: src/ngx_http_encrypted_session_cipher.c [hunk 1]
```diff
@@ -39,7 +39,7 @@ ngx_http_encrypted_session_aes_mac_encrypt(ngx_pool_t *pool, ngx_log_t *log,
         return NGX_ERROR;
     }
 
-    EVP_CIPHER_CTX_init(&ctx);
+    ctx = EVP_CIPHER_CTX_new();
```
Comment (agentzh) [inline src/ngx_http_encrypted_session_cipher.c]: This call may fail. Better check if `ctx` is NULL?

Comment (agentzh) [reply src/ngx_http_encrypted_session_cipher.c]: Also, can we reuse this allocated `ctx` data across different invocations? It would be great if we can save the expensive dynamic allocation on such a hot code path.

**Comment by spacewander** [comment]
@agentzh 
A new commit is pushed.

File: src/ngx_http_encrypted_session_cipher.c [hunk 2]
```diff
@@ -19,14 +19,23 @@
 static uint64_t ngx_http_encrypted_session_ntohll(uint64_t n);
 static uint64_t ngx_http_encrypted_session_htonll(uint64_t n);
 
+typedef int (*cipher_ctx_reset_handle) (EVP_CIPHER_CTX *ctx);
+
+#if OPENSSL_VERSION_NUMBER >= 0x1010000fL
+static cipher_ctx_reset_handle reset_cipher_ctx = EVP_CIPHER_CTX_reset;
+#else
+static cipher_ctx_reset_handle reset_cipher_ctx = EVP_CIPHER_CTX_cleanup;
```
Comment (agentzh) [inline src/ngx_http_encrypted_session_cipher.c]: Can we put these into the main conf and register a pool cleanup handle for it instead? Otherwise it's not valgrind clean and may have races upon HUP reload, maybe?

File: src/ngx_http_encrypted_session_cipher.c [hunk 3]
```diff
@@ -19,14 +19,23 @@
...
+#endif
+
+static EVP_CIPHER_CTX *session_ctx;
```
Comment (agentzh) [inline src/ngx_http_encrypted_session_cipher.c]: Also, file-level names should start with the `ngx_http_encrypted_session_` prefix.

**Comment by spacewander** [comment]
@agentzh 
Updated.

File: src/ngx_http_encrypted_session_module.c [hunk 4]
```diff
@@ -316,6 +326,70 @@ ngx_http_encrypted_session_expires(ngx_conf_t *cf, ngx_command_t *cmd,
 }
 
 
+static void
+ngx_http_encrypted_session_free_cipher_ctx(void *data)
+{
+    ngx_http_encrypted_session_main_conf_t      *esmcf;
```
Comment (agentzh) [inline src/ngx_http_encrypted_session_module.c]: We can just write `xxx     *esmcf = data;` here according to nginx's convention :)

File: src/ngx_http_encrypted_session_module.c [hunk 5]
```diff
@@ -316,6 +326,70 @@ ngx_http_encrypted_session_expires(ngx_conf_t *cf, ngx_command_t *cmd,
...
+        esmcf->session_ctx = NULL;
+    }
+}
+
+
+static ngx_int_t
+ngx_http_encrypted_session_init(ngx_conf_t *cf)
+{
+    ngx_pool_cleanup_t                        *cln;
+    ngx_http_encrypted_session_main_conf_t    *esmcf;
```
Comment (agentzh) [inline src/ngx_http_encrypted_session_module.c]: Maybe `emcf` is a better name here?

File: util/build.sh [hunk 6]
```diff
@@ -11,8 +11,8 @@ force=$2
 
 ngx-build $force $version \
             --with-http_ssl_module \
-            --with-cc-opt="-I$OPENSSL_INC -I$PCRE_INC" \
-            --with-ld-opt="-L$OPENSSL_LIB -L$PCRE_LIB -Wl,-rpath,$OPENSSL_LIB:$PCRE_LIB:$LIBDRIZZLE_LIB" \
+            --with-cc-opt="-I$OPENSSL_INC" \
+            --with-ld-opt="-L$OPENSSL_LIB -Wl,-rpath,$OPENSSL_LIB" \
```
Comment (agentzh) [inline util/build.sh]: Why this change? It looks like an accident check in.

Comment (spacewander) [reply util/build.sh]: @agentzh 
This build script will fail if we provide `OPENSSL_LIB` but not `$PCRE_LIB` and `$LIBDRIZZLE_LIB`. Since we only care about OpenSSL, I modify this script to let the build pass.

**Comment by spacewander** [comment]
@agentzh 
Updated.

File: util/build.sh (hunk 6 above)
Comment (agentzh) [reply util/build.sh]: @spacewander I think we should always support specifying our own PCRE. We don't want to use the system PCRE anyway. Better revert this.

Comment (spacewander) [reply util/build.sh]: @agentzh 
PCRE and libdrizzle are added in the travis build via a new commit now.

**Comment by agentzh** [comment]
@spacewander libdrizzle can be indeed removed since we do not even link against it :)

**Comment by spacewander** [comment]
@agentzh 
All right, libdrizzle is removed in the latest commit now.

**Comment by agentzh** [comment]
@spacewander Merged. Thanks!

**Comment by TheRusskiy** [comment]
@agentzh when is the new release?

**Comment by agentzh** [comment]
@TheRusskiy Maybe this month or next month.

**Comment by starriv** [comment]
Hi is coming May. When is the update version?

**Comment by agentzh** [comment]
A new release tag of this module was already added a few days ago and the new version is already included in OpenResty 1.13.6.2 RC1: https://groups.google.com/d/msg/openresty-en/bTfqTjshcX4/uP8qF0EpAgAJ
